shai/src/main/scala/at/forsyte/apalache/shai/TransExplorerService.scala

```diff
@@ -40,10 +68,75 @@ class TransExplorerService(connections: Ref[Map[UUID, Conn]]) extends ZioTransEx
    * @param req
    *   the request (isomorphic to the Unit)
    */
-  def openConnection(req: ConnectRequest): ZIO[ZEnv, Status, Connection] = for {
+  def openConnection(req: ConnectRequest): Result[Connection] = for {
     id <- ZIO.effectTotal(UUID.randomUUID())
     _ <- addConnection(Conn(id))
   } yield Connection(id.toString())
 
-  private def addConnection(c: Conn): UIO[Unit] = connections.update(_ + (c.id -> c))
+  /**
+   * Parses a spec into a model
+   *
+   * This method handles the [LoadModelRequest] RPC defined by `transExplorer.proto`
+   *
+   * @param req
+   *   the request to load a model, including the root module spec and any auxiliary modules
+   */
+  def loadModel(req: LoadModelRequest): Result[LoadModelResponse] = for {
+    parseResult <- parseSpec(req.spec, req.aux)
+    result <- parseResult match {
+      case Right(module) =>
+        for {
+          _ <- updateConnection(req.conn)(_.setModel(module))
```

Comment:
We are adding a connection to a map, which will replace the previous value of the key with the new value for that key. I.e., in 

```scala
  private def updateConnection(reqConn: RequestConn)(f: Conn => Conn): Result[Conn] = for {
    conn <- getConnection(reqConn)
    updatedConn = f(conn)
    _ <- addConnection(updatedConn)
  } yield updatedConn
```

We get from the map, then update the value, then set the new value by adding the connection back in.

The only thing that affects mutable state here is where we run the `update` method on the `connections` ref:

```scala
  private def addConnection(c: Conn): Result[Unit] = connections.update(_ + (c.id -> c))
```

If you think of any suggestions to improve the code here, I'd be happy to hear them! Also happy to sync up quickly to discuss :)

Reply:
Uff, I missed that the updated connection will have the same id.
Can we rename `addConnection` to `addOrUpdateConnection` then?
Also, `updateConnection` only works as intended if its parameter `f` does not change the connection id (otherwise it would *add* a connection)?

Reply:
How about changing it to `setConnection`? Would this work for you, since "setting" often the has the sense of adding something new if not already there? It doesn't do an "update" in the sense that `updateConnextion` does.

Good point on the possible misuse of the update function. I'll add a requirement on the `updateConenction` to enforce a condition that the id of the connection doesn't change.